WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift

```diff
@@ -64,100 +92,199 @@ struct WooShippingAddPackageView: View {
+                                }, label: {
+                                    Image(systemName: "chevron.backward")
+                                })
+                                .disabled(focusedField == WooShippingAddPackageDimensionView.DimensionType.allCases.first)
+                                Button(action: {
+                                    onForwardButtonTapped()
+                                }, label: {
+                                    Image(systemName: "chevron.forward")
+                                })
+                                .disabled(focusedField == WooShippingAddPackageDimensionView.DimensionType.allCases.last)
+                                Spacer()
+                                Button {
+                                    dismissKeyboard()
+                                } label: {
+                                    Text(Localization.keyboardDoneButton)
+                                        .bold()
+                                }
+                            }
+                            .renderedIf(focusedField != nil)
                         }
                     }
+                    Toggle(isOn: $showSaveTemplate) {
+                        Text(Localization.saveNewPackageTemplate)
+                            .font(.subheadline)
+                    }
+                    .tint(Color(.withColorStudio(.wooCommercePurple, shade: .shade60)))
+                    if showSaveTemplate {
+                        TextField("Enter a unique package name", text: $packageTemplateName)
+                            .font(.body)
+                            .focused($packageTemplateNameFieldFocused)
+                            .padding()
+                            .roundedBorder(cornerRadius: 8,
+                                           lineColor: packageTemplateNameFieldFocused ? Color(UIColor.wooCommercePurple(.shade60)) : Color(.separator),
+                                           lineWidth: packageTemplateNameFieldFocused ? 2 : 1)
+                        Button(Localization.savePackageTemplate) {
+                            savePackageAsTemplateButtonTapped()
+                        }
+                        .disabled(!validateCustomPackageInputFields())
+                        .buttonStyle(SecondaryButtonStyle())
+                        .padding(.bottom)
+                        .id(Constants.saveTemplateButtonID) // Set the id for the button so we can scroll to it
+                    }
                 }
-            } label: {
-                HStack {
-                    // text
-                    Text(packageType.name)
-                        .bodyStyle()
-                    // arrows
-                    Spacer()
-                    Image(systemName: "chevron.up.chevron.down")
+                .padding(.horizontal)
+                .onChange(of: showSaveTemplate) { newValue in
+                    packageTemplateNameFieldFocused = newValue
```

Comment:
Do you think we should also reset `packageTemplateName` to an empty string when this is toggled? It feels a little odd that any text entered there is retained if I toggle this off and on again.

Reply:
I can reset it but I am thinking what would be the least bad case:
- user enters a name, accidentally toggles it off, toggles it back on and the text is cleared (user needs to enter text again)
- user enters a name, toggles it off on purpose, toggles it back on and sees the last entered text (user can save with that name or change it) - **my preference**

I would prefer that the string is kept until the template is **saved**, and then we reset it. But we can see when we test it with more people what feels better?

Reply:
Yeah, good point, that's a fair concern.

> But we can see when we test it with more people what feels better?

Sounds good!